mediation_service/mediation/prepare_ssp_response.py

```diff
@@ -0,0 +1,183 @@
+        )
+        coding.code = "DATA_IN_TRANSIT"
+        coding_list.append(coding)
+        codeable_concept.coding = coding_list
+        codeable_concept.display = "Data in Transit"
+    if extension["valueCode"] == "data-awaiting-filing":
+        op_outcome_issue.diagnostics = (
+            "Patient data may be incomplete as there is data supplied by a third party "
+            "awaiting review before becoming available. "
+        )
+        coding.code = "DATA_AWAITING_FILING"
+        coding_list.append(coding)
+        codeable_concept.coding = coding_list
+        codeable_concept.display = "Data Awaiting Filing"
+
+    op_outcome_issue.details = codeable_concept
+
+    location_list = ["/entry"]
+    op_outcome_issue.location = location_list
+
+    return op_outcome_issue
+
+
+def __build_operationoutcome(operation_outcome_list) -> OperationOutcome:
+    operationoutcome = OperationOutcome()
+    meta = Meta()
+    fhir_date = FHIRDate()
+
+    fhir_date.date = datetime.datetime.now()
+    meta.lastUpdated = fhir_date
+    operationoutcome.meta = meta
+    operationoutcome.issue = operation_outcome_list
+
+    return operationoutcome
+
+
+def __transform_patient(ref, ssp_response: dict):
+    if not isinstance(ref, str) or "Patient" not in ref:
+        return ""
+
+    id_matches = re.findall(r"^Patient/(\d)", ref)
+    if id_matches:
+        patient_id = id_matches[0]
+        query = parse("`this`.entry[*].resource.resourceType")
+        matches = query.find(ssp_response)
+        for match in matches:
+            if match.value == "Patient":
+                index = match.full_path.left.left.right.index
+                patient = ssp_response["entry"][index]
+                if str(patient["resource"].get("id")) != patient_id:
+                    continue  # This is the correct patient only if it's 'id' matches with the one in ref
+
+                query = parse("resource.identifier")
+                matches = query.find(patient)
+                for identifier in matches:
+                    if (
+                        identifier.value.get("system").lower()
+                        == "https://fhir.nhs.uk/Id/nhs-number".lower()
+                    ):
+                        nhs_number = identifier.value.get("value")
+                        return f"AllergyIntolerance?patient:identifier=https://fhir.nhs.uk/Id/nhs-number|{nhs_number}"
+
+        return ""
+
+
+def __transform_local_references(ssp_response: dict):
+    query = parse("`this`.entry[*].resource.resourceType")
+    matches = query.find(ssp_response)
+
+    for match in matches:
+        if match.value == "AllergyIntolerance":
+            index = match.full_path.left.left.right.index
+            allergy = ssp_response["entry"][index]
+            query = parse("resource.patient.reference")
+            matches = query.find(allergy)
+            for ref in matches:
+                abs_ref = __transform_patient(ref.value, ssp_response)
```

Comment:
possible refactor - instead of parsing for Patients every time we have an Allergy Intolerance resource.
Parse for Patients before hand, generate dict mapping id to absolute ref.
Then for each Allergy tolerance grab correct ref from the dict?

Suggestion but up for discussion